crud/common/sharding_key.lua

```diff
@@ -0,0 +1,209 @@
+
+local ShardingKeyError = errors.new_class("ShardingKeyError", {capture_stack = false})
+local FetchShardingKeyError = errors.new_class('FetchShardingKeyError', {capture_stack = false})
+local GetShardingKeyFieldnosError = errors.new_class('GetShardingKeyFieldnosError',  {capture_stack = false})
+
+local FETCH_FUNC_NAME = '_crud.fetch_on_storage'
+
+local sharding_key_module = {}
+
+-- Build a structure similar to index, but it is not a real index object,
+-- it contains only parts key with fieldno's.
+function sharding_key_module.as_index_object(space_format, space_name, sharding_key_def)
+    dev_checks('table', 'string', 'table')
+
+    local fieldnos = {}
+    local fieldno_map = utils.get_format_fieldno_map(space_format)
+    for _, field_name in ipairs(sharding_key_def) do
+        local fieldno = fieldno_map[field_name]
+        if fieldno == nil then
+            return nil, GetShardingKeyFieldnosError:new(
+                "No such field (%s) in a space format (%s)", field_name, space_name)
+        end
+        table.insert(fieldnos, {fieldno = fieldno})
+    end
+
+    return {parts = fieldnos}
+end
+
+-- It's a public method to access it in tests.
+function sharding_key_module.fetch_on_storage()
+    local sharding_key_def = {}
+    local sharding_key_space = box.space._ddl_sharding_key
+    if sharding_key_space == nil then
+        return sharding_key_def
+    end
+
+    local SPACE_NAME_FIELDNO = 1
+    local SPACE_SHARDING_KEY_FIELDNO = 2
+    for _, tuple in ipairs(sharding_key_space:select()) do
+        local space_name = tuple[SPACE_NAME_FIELDNO]
+        local key = tuple[SPACE_SHARDING_KEY_FIELDNO]
+        sharding_key_def[space_name] = key
+    end
+
+    return sharding_key_def
+end
+
+local sharding_key_fetch_in_progress = false
+local sharding_key_fetch_cond = fiber.cond()
+
+-- Under high load we may get a case when many fibers will fetch sharding keys
+-- from storages. It is not good from performance point of view.
+-- concurrent_fetch() wraps a fetch_on_storage() to limit a number of fibers
+-- that fetches a sharding key by a single one, other fibers will wait for a
+-- FETCH_SHARDING_KEY_TIMEOUT and then make one more attempt to fetch sharding
+-- key if a cache is still empty.
+local function concurrent_fetch(replicaset)
+    dev_checks('table')
+
+    if sharding_key_fetch_in_progress == true and
+       not sharding_key_fetch_cond:wait(const.FETCH_SHARDING_KEY_TIMEOUT) then
+        return nil, FetchShardingKeyError:new(
+                        'Waiting for sharding key to be fetched is timed out')
+    end
+
+    local sharding_key_def = {}
+    if sharding_key_fetch_in_progress == false then
+        sharding_key_fetch_in_progress = true
+        local err
+        sharding_key_def, err = replicaset:call(FETCH_FUNC_NAME, {},
+                                                {timeout = const.FETCH_SHARDING_KEY_TIMEOUT})
+        if err ~= nil then
+            return nil, err
```

Comment:
Here you need to remove a lock. Otherwise in case of error it will be permanent lock.

Reply:
You are right, thanks. Fixed.
```lua
--- a/crud/common/sharding_key.lua
+++ b/crud/common/sharding_key.lua
@@ -78,6 +78,7 @@ local function concurrent_fetch(replicaset)
         sharding_key_def, err = replicaset:call(FETCH_FUNC_NAME, {},
                                                 {timeout = const.FETCH_SHARDING_KEY_TIMEOUT})
         if err ~= nil then
+            sharding_key_fetch_in_progress = false
             return nil, err
         end
         sharding_key_fetch_cond:broadcast()
```

Reply:
I don't see such code now and, it seems, we miss unlocking here.

Reply:
(And `:broadcast()` is missed in the patch above.)

Reply:
I'll close this comment, we decided to use `locked()`-like synchronization and `:broadcast()` is not needed anymore.
Open discussion is here - https://github.com/tarantool/crud/pull/181#discussion_r713273871

Reply:
Okay. We'll need to accurately look at failure cases (don't throw an error unexpectedly) and everything should be okay.